### ISSUE TITLE
Running many different parameters at a time

### ISSUE THREAD

**Issue by alessandramaranca** [niznik-dev/predicting-zygosity#62]
# Describe the Feature

Figuring out the right way to go through many parameters in data generation, fine-tune models on that data, and evaluate it (and output a csv of eval results by the iterated values - we need to add parameters to the evaluation in order to be able to add these to one single file with that specific p and N (for example). 

Either integrating it with our current slurm script generator or a separate script, but make sure that this is working so that we can leave code running for a long time testing many different things. Integrating ways to change our most common hyper parameters (for instance, number of epochs, learning rate, or the type of the model) might be great additions. 

# Acceptance Criteria
- Being able to go through many different data generating parameters (for instance, in generate.py, N, bit_length, etc) 

This is an example of a slurm script that goes through many different values of p (this I got working)

```
#!/bin/bash
#SBATCH --job-name=p-sweep
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --array=0-10
#SBATCH --gres=gpu:1
#SBATCH --mem=30G
#SBATCH --time=10:00:00
#SBATCH --mail-type=BEGIN,END
#SBATCH --mail-user=ar0241@princeton.edu
#SBATCH --output=logs/p-sweep_%a.out
#SBATCH --error=logs/p-sweep_%a.err

module purge
module load anaconda3/2024.6
conda deactivate 
conda deactivate 
conda activate torchtune 

# the sweep over p
p_list=(0 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0)
p=${p_list[$SLURM_ARRAY_TASK_ID]}

# per‐run directory
RUNDIR=/home/ar0241/scratch/twins/run_p${p}
mkdir -p "${RUNDIR}"

# 1) generate train.json & test.json
python scratch/twins/predicting-zygosity/tests/bit_sequences/generate.py \
  --bit_length 7 \
  --p "${p}" \
  --test_size 1000 \
  --outdir "${RUNDIR}" \
  --bit_parity True \
  --N 100000

# 2) finetune on the generated train.json
tune run lora_finetune_single_device \
  --config scratch/stochastic/train_test.yaml \
  epochs=10 \
  input_dir="${RUNDIR}" \
  output_dir="/home/ar0241/scratch/twins/output_p${p}"

# 3) evaluate on the generated test.json
export RESULTS_CSV=/home/ar0241/scratch/twins/results.csv
python scratch/stochastic/eval-3.py \
  --base_model_path /home/ar0241/scratch/torchtune_models/Llama-3.2-1B-Instruct \
  --adapter_path  /home/ar0241/scratch/twins/output_p${p}/epoch_9 \
  --eval_file      "${RUNDIR}/test.json" \
  --true_p         "${p}" \
  --results_csv   /home/ar0241/scratch/twins/results.csv


python scratch/stochastic/end-plot.py --N 100000   --bit_length 7   --test_size 10000   --eval_on_test false   --epochs 10   --results_csv /home/ar0241/scratch/twins/results.csv   --output_png /home/ar0241/scratch/twins/accuracy_vs_p.png
```

And this is an example of the type of desired further iteration. 

```
#!/bin/bash
#SBATCH --job-name=p-sweep
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --array=0-10
#SBATCH --gres=gpu:1
#SBATCH --mem=30G
#SBATCH --time=10:00:00
#SBATCH --mail-type=BEGIN,END
#SBATCH --mail-user=ar0241@princeton.edu
#SBATCH --output=logs/p-sweep_%a.out
#SBATCH --error=logs/p-sweep_%a.err

module purge
module load anaconda3/2024.6
conda deactivate
conda deactivate
conda activate torchtune

# --------------------------
# Parameters shared across runs
# --------------------------
epochs=10
test_size=1000
results_csv="/home/ar0241/scratch/twins/results.csv"

# Only task 0 deletes results.csv (to avoid race conditions)
if [[ "$SLURM_ARRAY_TASK_ID" == "0" ]]; then
  rm -f "$results_csv"
  echo "p,N,bit_length,accuracy" > "$results_csv"
fi

# --------------------------
# Sweep settings
# --------------------------
p_list=(0 0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1.0)
p=${p_list[$SLURM_ARRAY_TASK_ID]}

# Define sweep grid
N_values=(10000 100000 1000000)
bit_lengths=(5 10 15)

# Round-robin pairing of N and bit_length
N=${N_values[$((SLURM_ARRAY_TASK_ID % 3))]}
bit_length=${bit_lengths[$((SLURM_ARRAY_TASK_ID % 3))]}

# Per-run directory
RUNDIR="/home/ar0241/scratch/twins/run_p${p}_N${N}_bl${bit_length}"
mkdir -p "$RUNDIR"

# 1) Generate data
python scratch/twins/predicting-zygosity/tests/bit_sequences/generate.py \
  --bit_length "$bit_length" \
  --p "$p" \
  --test_size "$test_size" \
  --outdir "$RUNDIR" \
  --bit_parity True \
  --N "$N"

# 2) Finetune
tune run lora_finetune_single_device \
  --config scratch/stochastic/train_test.yaml \
  epochs="$epochs" \
  input_dir="$RUNDIR" \
  output_dir="/home/ar0241/scratch/twins/output_p${p}_N${N}_bl${bit_length}"

# 3) Evaluate
python scratch/stochastic/eval-3.py \
  --base_model_path /home/ar0241/scratch/torchtune_models/Llama-3.2-1B-Instruct \
  --adapter_path  /home/ar0241/scratch/twins/output_p${p}_N${N}_bl${bit_length}/epoch_9 \
  --eval_file     "${RUNDIR}/test.json" \
  --true_p        "$p" \
  --N             "$N" \
  --bit_length    "$bit_length" \
  --results_csv   "$results_csv"

```

**Comment by niznik-dev** [comment]
@alessandramaranca How do you feel about this input format? I chose yaml to be similar to what we already have to deal with:

```
outside_variables:
  p_value: [0, 0.25, 0.5, 0.75, 1]
  N: [1e4, 1e5, 1e6]
  BitLen: [1, 4, 7, 10]
torchtune_variables:
  epochs: [1, 2, 3, 4, 5]
  lr_scheduler/component:
     [
        'torchtune.training.lr_schedulers.get_cosine_schedule_with_warmup',
        'torchtune.training.lr_schedulers.get_linear_schedule_with_warmup',
        'torchtune.training.lr_schedulers.get_constant_schedule_with_warmup',
        'torchtune.training.lr_schedulers.get_exponential_schedule_with_warmup'
      ]
  optimizer/lr: [1e-4, 5e-5, 1e-5]
  model/lora_rank: [8, 16, 32, 64]
  batch: [1, 6, 11, 16]
```

**Comment by niznik-dev** [comment]
@alessandramaranca Based on what I worked on today, what will happen is each run will get its own output folder (something along the lines of runname_40G_5 inside of which should be a copy of the yaml used in the run which will contain all the parameters (including those torchtune doesn't care about but prepended with hsweeper_). I think that's as much as we'll need - just need to figure out how to trigger the eval to run on its own. Let's talk soon.
